pkg/services/ngalert/store/alert_rule.go

```diff
@@ -75,8 +69,72 @@ func (st DBstore) DeleteAlertRulesByUID(ctx context.Context, orgID int64, ruleUI
 		}
 		logger.Debug("Deleted alert rule state", "count", rows)
 
+		versions, err := st.getLatestVersionOfRulesByUID(ctx, orgID, ruleUID)
+		if err != nil {
+			logger.Error("Failed to get latest version of deleted alert rules. The recovery will not be possible", "error", err)
+		}
+		for idx := range versions {
+			version := &versions[idx]
+			version.ID = 0
+			version.RuleUID = ""
+			version.Created = TimeNow()
+			version.CreatedBy = nil
+			if user != nil {
+				version.CreatedBy = util.Pointer(string(*user))
+			}
+		}
+
+		rows, err = sess.Table(alertRuleVersion{}).Where("rule_org_id = ?", orgID).In("rule_uid", ruleUID).Delete(alertRule{})
+		if err != nil {
+			return err
+		}
+		logger.Debug("Deleted alert rule versions", "count", rows)
+
+		_, err = sess.Insert(versions)
+		if err != nil {
+			logger.Error("Failed to insert latest version of deleted alert rules. The recovery will not be possible", "error", err)
```

Comment:
I understand the intention not to fail when deleting the rule, but this might be unclear for the users in the UI when they expect all rules to be recoverable, and because of errors, they are not, and it's not visible. Maybe it'd be better to fail here and return an error?

Reply:
I agree. However, I am not sure if there is a way to solve this error if it happens.

Reply:
I decided to put it behind feature flag (which actually we intended to do anyway)